### PR TITLE
Add stream and memory-resource parameters to struct-scalar copy ctor

### PR DESCRIPTION
**Pull request by davidwendt**
Investigating issue #8883 I found the `cudf::struct_scalar` copy constructor does not take a stream or memory resource parameter like the other scalar copy constructors. This PR adds the appropriate signature and implementation.

Also found usage of `thrust::device_pointer_cast` being used in strings concatenate specialization to get values from an offsets column. This PR changed those to use `cudf::detail::get_value` instead to honor the current stream.

### CODE REVIEW

**Review by vyasr — APPROVED** [review]
Nice simplification of the offsets calc, and +1 for removing includes from the tests.

File: cpp/src/strings/copying/concatenate.cu [hunk 1]
```diff
@@ -287,12 +285,15 @@ std::unique_ptr<column> concatenate(host_span<column_view const> columns,
...
+        auto bytes_offset =
+          cudf::detail::get_value<offset_type>(offsets_child, column_offset, stream);
 
         // copy the chars column data
-        auto d_chars    = chars_child.data<char>() + bytes_offset;
-        size_type bytes = thrust::device_pointer_cast(d_offsets)[column_size] - bytes_offset;
+        auto d_chars = chars_child.data<char>() + bytes_offset;
+        auto const bytes =
+          cudf::detail::get_value<offset_type>(offsets_child, column_size + column_offset, stream) -
+          bytes_offset;
```
Comment (karthikeyann) [inline cpp/src/strings/copying/concatenate.cu:295]: This part of the code, overlooks the size of `chars_child` column. while debugging a memory error, I found that `chars_column.size()` might be less than `bytes` and it caused `cudaMemcpyAsync` invalid argument failure.

Probably a `CUDA_EXPECTS` might satisfy size checking.

Comment (davidwendt) [reply cpp/src/strings/copying/concatenate.cu:295]: It is easy enough to add the `CUDF_EXPECTS` but the `cudaMemcpyAsync` is wrapped in a `CUDA_TRY` so an exception is thrown anyway.

**Review by karthikeyann — APPROVED** [review]

**Comment by davidwendt** [comment]
@gpucibot merge

**Pull request merged**